lib/src/package.dart

```diff
@@ -131,22 +131,48 @@ class Package {
   ///
   /// `pubspec_overrides.yaml` is only loaded if [withPubspecOverrides] is
   /// `true`.
+  ///
+  /// [alreadyLoadedPubspecs] contains a map from the `p.canonicalize`d
+  /// directory containing a pubspec, to the loaded [Pubspec] object.
+  ///
+  /// This mechanism can be used to avoid loading pubspecs twice. It can also be
+  /// used to override a pubspec in memory for trying out an alternative
+  /// resolution.
   factory Package.load(
     String? name,
     String dir,
     SourceRegistry sources, {
     bool withPubspecOverrides = false,
+    Map<String, Pubspec> alreadyLoadedPubspecs = const {},
```

Comment:
Instead of this, how about passing a `loadPubspec()` function?

Then you won't need to document the structure of the map, and the fact that things are cached is something you can hide altogether.

Reply:
Yeah - good idea!

Reply:
Done